nixos/modules/virtualisation/containers.nix

```diff
@@ -29,12 +29,30 @@ in
     };
 
     cdi = {
-      dynamic.nvidia.enable = mkOption {
-        type = types.bool;
-        default = false;
-        description = lib.mdDoc ''
-          Enable dynamic CDI configuration for NVidia devices by running nvidia-container-toolkit on boot.
-        '';
+      dynamic.nvidia = {
+        enable = mkOption {
+          type = types.bool;
+          default = false;
+          description = lib.mdDoc ''
+            Enable dynamic CDI configuration for NVidia devices by running nvidia-container-toolkit on boot.
+          '';
+        };
+
+        mount-nvidia-binaries = mkOption {
+          default = true;
+          type = types.bool;
+          description = lib.mdDoc ''
+            Mount binaries nvidia-smi, nvidia-cuda-mps-control, nvidia-cuda-mps-server, nvidia-debugdump, nvidia-powerd and nvidia-ctk on containers.
```

Comment:
This isn't exact though? I think we always mount them but normally we'd mount the store path and expose them in `PATH`. What we need an explicit user's consent for is for contaminating `/usr/bin`

Reply:
Without this setting, those executables will no tbe mounted at all on the container namespace unless I have overlooked something; I cannot find them being mounted directly nor indirectly within the container.

Reply:
Mhm. I think the right thing to do is to mount `/nix/store/.../bin/nvidia-smi=/nix/store/.../bin/nvidia-smi` and to extend `PATH`. AFAIU `containerEdits.env` lets one do that?

Reply:
Unfortunately we cannot do that. You can set or override envvars, but cannot extend an existing one. It's not possible to do `PATH=something-new:$PATH` so to say on the OCI hook.

However, I believe the original comment was about the fact of mounting these executables and my comment was to confirm that if you disable this option, no executables will be mounted at all.

Reply:
> Unfortunately we cannot do that. You can set or override envvars, but cannot extend an existing one. It's not possible to do PATH=something-new:$PATH so to say on the OCI hook.

That's alright, we have the module system at our disposal to take care of the module sematnics. And then our script can also graciously handle the potential values `nvidia-ctk` could have set before.

Reply:
1. We've implicitly already allocated the space for `virtualisation.containers.cdi.static."nvidia-container-toolkit".containerEdits.env`, but I'm not sure if we should make it an attrset on our side and convert to a list when passing to `cdi-generate.nix`
2. Maybe we make a separate option...

Reply:
I don't think we can do much with `PATH`. Users might have built container images with a specific `PATH` and we should not override it, given we cannot expand it.

Reply:
Offtopic: I'm still wondering if `cdi.static` and `cdi.dynamic could have been `cdi.nvidia-container-toolkit.settings` and `cdi.nvidia-container-toolkit` respectively.

Reply:
Hmmm, populating `PATH` may still be an issue when we deal with multiple devices/multiple hooks

Reply:
> Offtopic: I'm still wondering if cdi.static and cdi.dynamic could have been cdi.nvidia-container-toolkit.settings and cdi.nvidia-container-toolkit respectively.

- `cdi.static` gives the generic ability to configure CDI for generic devices in a manual fashion, just making sure by the type system that what they provide is a valid syntactic JSON.

- `cdi.dynamic` is a bit the "counterpart", where providers (only nvidia for now) provide you with the generic ability to configure in runtime based on the vendor detected hardware.

`cdi.static` has nothing to do with `nvidia-container-toolkit` necessarily, so I would keep them separate.

Also, having `cdi.nvidia-container-toolkit` directly could be an option, but given there is a very good split on the CDI spec of static/dynamic and their relationship (priority dynamic > static), I thought it was a good strategy to map them in that way in our options.

Reply:
Note on that: https://github.com/cncf-tags/container-device-interface/blob/780f60f5e4ac27d7b1c729a986ed7edbb6af7c12/pkg/cdi/doc.go#L157-L164

Reply:
> cdi.static gives the generic ability to configure
> cdi.dynamic is a bit the "counterpart",
> cdi.static has nothing to do with nvidia-container-toolkit necessarily, so I would keep them separate.

Yes, you had explained the logic in the original PR. However, we do implicitly maintain both "static" and "dynamic" options for `/run/cdi/nvidia-container-toolkit.json`: all of our `mounts` are currently computed statically (although we really should wrap them into a build with `exportReferencesGraph`), but we also do run `nvidia-ctk cdi generate` which is dynamic. And we still have to take care of merging the dynamic part with the static overrides (currently via `jq`).

I'd rather there were exactly one root option for `/run/cdi/nvidia-container-toolkit.json` (e.g. `cdi.nvidia-container-toolkit`). It is only logical to describe the static part in the style of RFC 42 ("structural `settings`"). The dynamic part then consists of running arbitrary hooks (`nvidia-ctk cdi generate`), possibly expanding the static configuration (e.g. when `mounts` contain store paths they need to be passed through the `exportReferencesGraph`), and finally merging the two

Reply:
This is why I was suggesting that maybe we should remove `cdi.dynamic.nvidia` and keep `services.nvidia-ctk`(?):

```
options.virtualisation.containers.cdi = mkOption { type = attrsOf (submodule { options.containerEdits.env = mkOption { type = attrsOf str; }; ... }); ... } # Literally mirroring the full CDI spec, because we can
options.services.nvidia-container-toolkit.enable = ...
options.services.nvidia-container-toolkit.mount-usr-local-bin = ...
options.services.nvidia-container-toolkit.mount-usr-lib = ...
config = mkMerge [
  (mkIf config.services.nvidia-ctk.enable {
    # Corresponds to /run/cdi/nvidia-container-toolkit.json
    virtualisation.containers.cdi.nvidia-container-toolkit.containerEdits.mounts = ...
  })
]
```

etc

Reply:
> However, we do implicitly maintain both "static" and "dynamic" options for /run/cdi/nvidia-container-toolkit.json

Ideally we wouldn't have to add any mounts with `jq` on the dynamically generated JSON with `nvidia-container-toolkit cdi generate`. This is the easiest path to make this work. Ideally, we should contribute changes to  the nvidia-container-toolkit to allow for more customization when it makes sense for that project and reduce the number of patches we do on the generated JSON.

Despite we patch the generated JSON, everything we write into `/run/cdi` I consider it "dynamic" because is what goes into the "dynamically generated, although extended a bit by us to make it work fine". Without those patches, the JSON generated by the nvidia container toolkit under /var/run/cdi would not work on a NixOS host.

> I'd rather there were exactly one root option for /run/cdi/nvidia-container-toolkit.json (e.g. cdi.nvidia-container-toolkit)

I think it all boils down to the fact that you see `nvidia-container-toolkit-cdi-generator` as a module that should not exist, and instead be implemented in `virtualisation.containers` directly?

We can maybe get rid of this internal module (nvidia-container-toolkit-cdi-generator) and make the move if that is your main concern. I think it is a clean separation though.

Reply:
> Note on that: https://github.com/cncf-tags/container-device-interface/blob/780f60f5e4ac27d7b1c729a986ed7edbb6af7c12/pkg/cdi/doc.go#L157-L164

My impression is that this only concerns the generic placeholder static configs with default values that you distribute with packages in distributions like ubuntu. The "static config" we are discussing actually contain values specified by the end user, i.e. they are more relevant and informed than the values produced by `nvidia-ctk cdi generate` and should generally take the priority.

TLDR: the spec's distinction between "static" and "dynamic" so far seems fairly useless, especially given that it doesn't cover merge semantics

> Ideally, we should contribute changes to the nvidia-container-toolkit to allow for more customization when it makes sense for that project and reduce the number of patches we do on the generated JSON.

What I see so far suggests that we'll always want to apply "edits" to the configs generated by tools like "nvidia-ctk" (including those by other vendors). So I think we need a generic system for describing and applying the common overrides.

Reply:
> I think it all boils down to the fact that you see nvidia-container-toolkit-cdi-generator as a module that should not exist, and instead be implemented in virtualisation.containers directly?

Either that or the `cdi.dynamic.nvidia` shouldn't, because they're so entangled they just mirror each other.

BUt that's not the point. The point is that ultimately we're working on a module to manage files in `/run/cdi/`. Right now there's no generic formula to say which options correspond to which files.